### DISCUSSION TITLE
Continuous and hybrid custom constraints

### DISCUSSION THREAD

**Discussion by BMGla5er** [emdgroup/baybe#363] [Q&A]
I am interested in using custom constraints to subset the parameter space, using a surrogate NN to predict a target value and determine if it is above a given threshold. To give context, this will be a materials science question, where I provide as many as five element compositions to predict alloy structure, and use baybe to maximize performance. When constructing a custom constraint on a test case, I could not use a continuous parameter set, and when using a discrete set it takes a very long time to recommend the next location due to the large number of steps (my test case used three parameters in a range from 0 to 10, in steps of 0.1).

Will this function be enabled for continuous parameter sets in future updates, or do you have a recommendation for how to accelerate this?

**Comment by AdrianSosic** [comment]
Hi @BMGla5er, thanks for your interest in the package. Correct, at the moment, there is no such thing available as a `CustomContinuousConstraint` but it should require only minor changes in the code to enable this. I suspect we are talking about a nonlinear constraint? In that case, what you need to do is to implement your constraint in a way that is compatible with botorch's `nonlinear_inequality_constraint` argument to [`optimize_acqf`](https://github.com/pytorch/botorch/blob/c47c01ca7cad83c068efdfb24f30242ba5d01e6d/botorch/optim/optimize.py#L397), add that constraint to the search space via our [`constraints_nonlin`](https://github.com/emdgroup/baybe/blob/4f65c7e91e26def29108c0993e7ae0c83f51daae/baybe/searchspace/continuous.py#L65) argument, and add a corresponding line of code that passes the constraint to the `optimize_acqf` [call](https://github.com/emdgroup/baybe/blob/4f65c7e91e26def29108c0993e7ae0c83f51daae/baybe/recommenders/pure/bayesian/botorch.py#L158). I think we can also enable this in one of the next PRs, but it would be great to hear some of your experiences first.

Can you share some details on what the constraint is supposed to do?

**Comment by AdrianSosic** [comment]
Hi @BMGla5er, any more thoughts on this? Otherwise, I'll close the discussion this week 🙃

**Comment by Scienfitz** [comment]
closing as inactive
